go/interpreter/lfvm/memory.go

```diff
@@ -68,68 +69,75 @@ func (m *Memory) ExpansionCosts(size uint64) tosca.Gas {
 	return fee
 }
 
-func (m *Memory) EnsureCapacity(offset, size uint64, c *context) error {
-	if size <= 0 {
-		return nil
-	}
+// expandMemoryAndCharge tries to expand memory to the given size.
+// If the memory is already large enough or size is 0, it does nothing.
+// if not enough gas in the context, it returns an error.
+func (m *Memory) expandMemoryAndCharge(offset, size uint64, c *context) error {
```

Comment:
Instead of using closures and callbacks this could be simpler implemented by one function calling the other:
```
func (m *Memory) expandMemoryAndCharge(offset, size uint64, c *context) error {
	if size == 0 {
		return nil
	}

	needed := offset + size
	// check overflow
	if needed < offset {
		c.signalError()
		return errGasUintOverflow
	}

	fee := m.expansionCosts(needed)
	if !c.useGas(fee) {
		c.status = statusOutOfGas
		return errOutOfGas
	}
	m.total_memory_cost += fee

	m.expandMemory(needed)
	return nil
}

// expandMemory expands the memory to the given size without checking gas.
func (m *Memory) expandMemory(needed uint64) {
	needed = toValidMemorySize(needed)
	size := m.len()
	if size < needed {
		m.store = append(m.store, make([]byte, needed-size)...)
	}
}
```